Skycultures 2.0: Zodiac and Lunar Systems

gzotti opened the pull request:
### Description
This branch should introduce optional cultural elements: zodiac and various flavours of Lunar stations or mansions. The latter are important in many Asian cultures, while the Zodiac is best known from Mesopotamia to Europe. There are different definitions and variants to be found in the literature which we have to learn and may have to adjust and extend even after the PR will be regarded good for a first release.

Fixes #437, fixes #1627 

### Screenshots (if appropriate):

### Type of change
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] This change requires a documentation update
- [ ] Housekeeping

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

**Test Configuration**:
* Operating system: Win11
* Graphics Card: irrelevant

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [ ] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (header file)
- [x] I have updated the respective chapter in the Stellarium User Guide
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

gzotti commented:
This won't sound very difficult. However, is it possible that someone in 400BC would have tried to compute backward? First things first, this has to work, but we can add some "introduced" tag later.

sushoff commented on skycultures/modern/index.json:
```diff
@@ -1675,6 +1675,53 @@
+			 { "symbol": "\u264C", "native": "Leo", "english": "Lion"},
+			 { "symbol": "\u264D", "native": "Virgo", "english": "Virgin"},
+			 { "symbol": "\u264E", "native": "Libra", "english": "Scales"},
+			 { "symbol": "\u264F", "native": "Scorpio", "english": "Scorpion"},
```
the name is "Scorpius" (Latin, not Italian!)

gzotti replied:
I wanted to look a bit medieval here, whatever is commonly found on old prints. Can be reversed of course.

sushoff commented on skycultures/modern/index.json:
```diff
@@ -1675,6 +1675,53 @@
+			 { "symbol": "LS06", "native": "आर्द्रा",             "pronounce": "Ardra",      "english": "Storm God"},
+			 { "symbol": "LS07", "native": "पुनर्वसु ",            "pronounce": "Punarvasu",  "english": "Two Restorers of Goods"},
+			 { "symbol": "LS08", "native": "पुष्य",              "pronounce": "Pushya",    "english": "Nourisher"},
+			 { "symbol": "LS09", "native": "आश्ळेषा/आश्लेषा",       "pronounce": "Āshleshā",   "english": "Embrace"},
```
why does "Āshleshā" have diacritica but the other not?

gzotti replied:
I quickly copy/pasted the strings from Wikipedia. I don't pretend to read/write native Indian. At this point, in this branch while in draft, I need strings, and will gladly take corrections later. :-)

sushoff commented on skycultures/modern/index.json:
```diff
@@ -1675,6 +1675,53 @@
+			 { "symbol": "LS07", "native": "पुनर्वसु ",            "pronounce": "Punarvasu",  "english": "Two Restorers of Goods"},
+			 { "symbol": "LS08", "native": "पुष्य",              "pronounce": "Pushya",    "english": "Nourisher"},
+			 { "symbol": "LS09", "native": "आश्ळेषा/आश्लेषा",       "pronounce": "Āshleshā",   "english": "Embrace"},
+			 { "symbol": "LS10", "native": "मघा",              "pronounce": "Maghā",     "english": "Bountiful"},
```
please make it consistently (either all with or all without diacritics)

gzotti replied:
30 minutes vs. 6 hours. Yes, later.

sushoff replied:
https://xing.fmi.uni-jena.de/mediawiki/index.php/Indian_Names_(All_Terms)

gzotti replied:
To make it consistent, should I use the column with or without diacritics? Look in LS2, and a few others: You list totally different names in the columns. How do I know what's the correct transliteration of the Sanskrit? You know what: I let you fix the strings. :-) I moved the block into Indian SC now.

gzotti replied:
@vvasuki provided the strings now.

sushoff commented on skycultures/modern/index.json:
```diff
@@ -1675,6 +1675,53 @@
+			 { "symbol": "LS26", "native": "उत्तर भाद्रपदा/उत्तर प्रोष्ठपदा", "pronounce": "Uttara Bhādrapadā", "english": "Second of the Blessed Feet"},
+			 { "symbol": "LS27", "native": "रेवती",            "pronounce": "Revati",            "english": "Prosperous"}, 
+			 { "symbol": "LS27", "native": "अभिजित",          "pronounce": "Abhijit",           "english": "Invincible"}]},			 
   "edges_type": "iau",
```
really? 

In my view, as the nakshatras represent an own, ~4000 years old culture, they should not be mixed with modern boundaries

gzotti replied:
As said, this is a branch in draft. The Nakshatras will not stay in the modern SC, this is just to have the display immediately after launch. Obvious candidate for integration is Vedic, but you can create your isolated version as well. But let me get the implementation right before moving and fixing the json.

sushoff commented on src/core/StelSkyCultureSkyPartition.hpp:
```diff
@@ -0,0 +1,156 @@
+//! 	       { "symbol": "\u264C", "native": "Leo",         "english": "Lion"},
+//! 	       { "symbol": "\u264D", "native": "Virgo",       "english": "Virgin"},
+//! 	       { "symbol": "\u264E", "native": "Libra",       "english": "Scales"},
+//! 	       { "symbol": "\u264F", "native": "Scorpio",     "english": "Scorpion"},
```
dito: Scorpius

gzotti replied:
-"-

gzotti commented:
OK, we are getting close. GUI is there, but not totally working (fader setup? - and I don't get why Zodiac fade time is frozen at 10s.) The Chinese LM (currently only in Chinese SC) are taken from WP again, it seems they agree with the old Mansion "Borders/Edges" except for LM5. Given that probably all "defining stars" bear the number one in their respective constellation, this line (through a star II) may be wrong in WP, but it needs confirmation by SC experts. 
Is there any historical source explaining the vastly different mansion sizes? I see LM4 very narrow, LM8 quite wide. Does that have to do with Chin. numerology and lucky/unlucky numbers? What about LM20? Super-unlucky?

Currently the mansion lines are linked to stars in current epoch/equinox, lines drawn in equinox of date. Defining stars are marked with a little circle, that was a good idea!

Now the question remains, should the mansions really be set up on epochs, and then lose their defining stars, letting stars slip by precession etc? How would that have been observed? Or were they defined in a way that "from now on these stars define the LM", using celestial poles and equator at time of observation?

Atque commented:
The labels for Sagittarius and Capricorn are reversed:
![image](https://github.com/user-attachments/assets/30c04020-99c5-4114-9390-1f594575087d)

Atque commented:
Symbols need to be reversed too; Sagittarius = ♐︎, Capricornus = ♑︎. ;)

alex-w commented:
> OK, we are getting close. GUI is there, but not totally working (fader setup? - and I don't get why Zodiac fade time is frozen at 10s.) 

The result of quick looking the code for new features - the both faders run incorrectly. Please see using faders in grids and constellations classes for examples.

Atque commented:
Mirroring the view strangely prints the labels upside-down.
![image](https://github.com/user-attachments/assets/f66e7ddd-9330-4f23-8e3e-d5b8e7909440)

sushoff commented:
> OK, we are getting close. GUI is there, but not totally working (fader setup? - and I don't get why Zodiac fade time is frozen at 10s.) The Chinese LM (currently only in Chinese SC) are taken from WP again, it seems they agree with the old Mansion "Borders/Edges" except for LM5. Given that probably all "defining stars" bear the number one in their respective constellation,

by definition (this is the only star that was given with coordinates in hist. Chin. star cat.s 

> this line (through a star II) may be wrong in WP,

not necessarily: the counting has not been canonical prior to the Jesuits

What WP page did you use? You can cross check with our "Chinese Song Dyn." (the one which has a map in description) and "Chinese Chenzhuo" star catalog data. There is almost a millennium between these two historical documents - but both our SCs have the LM as "boundaries". Any potential difference between the two SCs is definitely in the historical source. 

> but it needs confirmation by SC experts. Is there any historical source explaining the vastly different mansion sizes? I see LM4 very narrow, LM8 quite wide.

is correct - see hist. map in our Chin. SC description. 

> Does that have to do with Chin. numerology and lucky/unlucky numbers? What about LM20? Super-unlucky?
> 
> Currently the mansion lines are linked to stars in current epoch/equinox, lines drawn in equinox of date. Defining stars are marked with a little circle, that was a good idea!

Thanks!
> 
> Now the question remains, should the mansions really be set up on epochs, and then lose their defining stars, letting stars slip by precession etc? 

no, I would still anchor the LM at stars (and move with precession and pm).  

> How would that have been observed?

armillary sphere and similar instruments ... there were many, but I would declare their adjustment a source of error and the definition at stars as "correct" / original/ meant to be ... 

> Or were they defined in a way that "from now on these stars define the LM", using celestial poles and equator at time of observation?

yes

sushoff commented:
something really IMPORTANT

- Scorpi**O** is Italian 
- Scorpi**US** is Latin 

The Latin term is used for the sign (ecliptic section) and for the constellation in astronomy. 

The bad habit of using the Italian term in popculture-astrology must not be distributed further in Stellarium!!!

gzotti commented:
> ### something really IMPORTANT
> 
>     * Scorpi**O** is Italian
> 
>     * Scorpi**US** is Latin
> 
> 
> The Latin term is used for the sign (ecliptic section) and for the constellation in astronomy.
> 
> The bad habit of using the Italian term in popculture-astrology must not be distributed further in Stellarium!!!

Actually, consulting my Latin dictionary, Scorpio, -onis is the primary Latin form (closer to Greek), and Scorpius, -ii is listed as "poetic variant". It has nothing to do with "Italian". In my (probably incomplete) searches I have the impression that the *zodiacal sign* is more often called "Scorpio" and described as "associated with the constellation Scorpius".

gzotti commented:
>>The Chinese LM (currently only in Chinese SC) are taken from WP again, it seems they agree with the old Mansion "Borders/Edges" except for LM5. Given that probably all "defining stars" bear the number one in their respective constellation,
> 
> by definition (this is the only star that was given with coordinates in hist. Chin. star cat.s
> 
> > this line (through a star II) may be wrong in WP,
> 
> not necessarily: the counting has not been canonical prior to the Jesuits
>
> 
> What WP page did you use? 

https://en.wikipedia.org/wiki/Twenty-Eight_Mansions

They list &alpha; Sco, you had defined the line with &sigma; Sco as found also elsewhere. I changed that now to &sigma;

>You can cross check with our "Chinese Song Dyn." (the one which has a map in description) and "Chinese Chenzhuo" star catalog data. There is almost a millennium between these two historical documents - but both our SCs have the LM as "boundaries". Any potential difference between the two SCs is definitely in the historical source.

I have added (copy/pasted) the (currently only single) description to the other Chinese SCs. A few more borders are different, (and currently the new definition goes through the star and equatorial poles of date, not of epoch, while the "abused borders" are frozen in epoch).


> > but it needs confirmation by SC experts. Is there any historical source explaining the vastly different mansion sizes? I see LM4 very narrow, LM8 quite wide.
> 
> is correct - see hist. map in our Chin. SC description.

Yes, but presumably there was a reason for this. There would be enough candidate stars for a more equal distribution. Proper motion or even only precession seems not to be sufficient as "width-changing" factor.  

> > Does that have to do with Chin. numerology and lucky/unlucky numbers? What about LM20? Super-unlucky?
> > Currently the mansion lines are linked to stars in current epoch/equinox, lines drawn in equinox of date. Defining stars are marked with a little circle, that was a good idea!
> 
> Thanks!
> 
> > Now the question remains, should the mansions really be set up on epochs, and then lose their defining stars, letting stars slip by precession etc?
> 
> no, I would still anchor the LM at stars (and move with precession and pm).
> 
> > How would that have been observed?
> 
> armillary sphere and similar instruments ... there were many, but I would declare their adjustment a source of error and the definition at stars as "correct" / original/ meant to be ...

Yes, I expect the same. With this you can easily observe equatorial coords of date, but not some precession-shifted lines after 300 years.

 
> > Or were they defined in a way that "from now on these stars define the LM", using celestial poles and equator at time of observation?
> 
> yes

That means, the current implementation of the new lines is OK (apart from declaring a few different defining stars), and the epoch-frozen old borders are no longer relevant?

sushoff commented:
I'll get back later ...

gzotti commented:
> @gzotti as we talked about the invention of nakshatras, zodiac etc. yesterday:
> 
> I know, it makes it more complicated but: shouldn't we add a quality like "visibility" for these mathematical concepts? For instance, if the zodiac was invented in 400 BCE, then it must not be displayed before 410 BCE because otherwise people would derive crazy things.

I am still not sure if it's a good idea, as going back at least a few centuries can help understanding the formation of the systems. If you really want that, please provide a list of "introduced" tags (years with decimal fraction) for:

- "tropical zodiac" (modern SCs)
- Indian lunar stations (nakshatras. What about the earlier 28-station system? [this may be defined later])
- Indian "sidereal" Zodiac (same period? later?)
- Chinese mansions. Probably a very concrete date can be given here, i.e. the maps that we try to replicate. 

I don't know about the Arab stations: are they just the asterisms, or also ecliptic sections that can be defined?

gzotti commented on src/gui/ViewDialog.cpp:
```diff
@@ -619,10 +619,10 @@ void ViewDialog::createDialogContent()
-		ui->constellationHullsColorButton->hide();
-		ui->constellationHullsThicknessSpinBox->hide();
-		ui->constellationHullsFadeDurationDoubleSpinBox->hide();
+                ui->constellationHullsCheckBox->setEnabled(false);
```
My intention really was to not even show these elements to beginning users. Those 20 (?) researchers interested in the convex hulls should enable the setting in config.ini. Showing but disabling make zero sense IMHO.

alex-w replied:
After adding UI elements for Zodiac and Lunar Systems the "Hulls" elements is looks like an "empty' place in UI - this is not very good

gzotti replied:
But we have other "holes" elsewhere in the GUI. Look into the SSO tab. I consider this "hole" here not so critical, and certainly better than filling it with visual but disabled clutter. I would like to just skip your commit on next rebase -i. 

We could think of moving const. hulls below const. boundaries, and zodiac and lunar stations to the right side. This would then have a "hole" in the lower left corner, maybe less disturbing for you. 

What should still come (and was on my task list of further refinements) is to check after switching a SC and (visually) disable zodiac or LS elements if a SC does not have it. Plus, exchanging the label by the actual culture-relevant name. But these are minor details, and can even wait for 25.3, when maybe all terms have been translated.

Speaking of translation, using just one user language setting instead of two is still open, and apparently some issues with missing context (Men, Leo, Ara, ...), so maybe some context rules should be established?

alex-w commented:
@gzotti what about English names for Zodiac? Currently not all Zodiac elements using English names as a synonyms of English names of IAU constellations. I think it should be synced (English names and context)

gzotti commented:
I had the similar discussion around the "native" spelling of the Scorpius/Scorpion above. I am very unsure whether any english-speaking person uses "Sea Goat" as name of the zodiacal sign commonly written as "Capricorn". The same is true for German "Steinbock". Nobody says "Ziegenfisch", unless you are about to explain the Babylonian origin of the figure in a planetarium show. We had user complaints already.

gzotti commented:
Yes, indeed, it seems necessary. Now, can context be formed from "skyculture + zodiac" and "skyculture + lunar_system" automatically, or do we need a "context" tag in the json ? This question may appear in several places now. Even, it appears, down to "skyculture+constellation abbreviation", "skyculture+constellation translated name", "skyculture+constellation pronunciation".

alex-w commented:
> Yes, indeed, it seems necessary. Now, can context be formed from "skyculture + zodiac" and "skyculture + lunar_system" automatically, or do we need a "context" tag in the json ? This question may appear in several places now. Even, it appears, down to "skyculture+constellation abbreviation", "skyculture+constellation translated name", "skyculture+constellation pronunciation".

It can be formed as will be coded in python code for extraction the translatable data. Right now all translatable data are extracted without context if "context" element is not defined in JSON. Of course we may add one context for whole SC in python and C++ code (Skyculture ID for example) and removed support special JSON elements "context". Or adding JSON elements "context" for all places where context is needed.

@10110111 @sushoff  please your opinion

10110111 commented:
> please your opinion

If context is required in >50% of entries, then I'd go with the fully automated solution. If it's just needed in <25% of entries, I'd prefer manual specification. The cases in between are a judgement call.

gzotti commented:
> our wizard (or djinn), Khalid, defined sections: see "Arabic lunar stations" SC

Ah, indeed. But I see several shortcomings after 25.1, maybe they can be improved (by name experts, you or Khalid?): 
- constellation names are Arab and translated (english) only, pronounce is missing. 
- numerical constellation abbreviations. Unless there is a good reason like referencing some external index, I recommend short character-based abbrevs.
- With 13/14 degrees wide lunar stations, this must again be coded as new variation. Is it just a list of ecliptical longitudes of J2000, and can this list be provided please (or provide an accessible reference)? Are there sub-divisions? Would it be OK to limit the zone along the ecliptic to a common width (e.g. &pm; 10 or 15°?)
- I recommend repeating the anwa-defining stick figures as asterisms to clearer highlight them.

Atque commented:
Zodiac + lunar mansions should probably only be visible from Earth (Zodiac maybe also from the Moon). I get crazy views from other planets.

alex-w reviewed (approved):
The new feature is OK for me, but it has problem for localization support (though it's not this PR issue)

gzotti commented:
Yes, I agree that after this is merged, the various shortcomings in translations that appeared with the new features in the last months should be solved, at the latest in 25.3. Some issues like the multiple-scripts-one-language issue (#4365) can hopefully be clarified in translation guidelines.

gzotti commented:
Argh, right. I started with this before I had the GUI and forgot to revert. Will fix ASAP.

alex-w commented:
> Argh, right. I started with this before I had the GUI and forgot to revert. Will fix ASAP.

I fixed it already
